crates/sui-core/src/signature_mpc/sign.rs

```diff
@@ -117,33 +142,76 @@ impl SignState {
+        public_nonce_encrypted_partial_signature_and_proofs: Vec<
+            PublicNonceEncryptedPartialSignatureAndProof<ProtocolContext>,
+        >,
+        presigns: Vec<DecentralizedPartyPresign>,
     ) {
         self.messages = Some(messages);
-        self.public_nonce_encrypted_partial_signature_and_proofs = Some(public_nonce_encrypted_partial_signature_and_proofs);
+        self.public_nonce_encrypted_partial_signature_and_proofs =
+            Some(public_nonce_encrypted_partial_signature_and_proofs);
+        self.presigns = Some(presigns);
     }
 
     pub(crate) fn insert_first_round(
         &mut self,
-        party_id: PartyID,
-        message: Vec<(PaillierModulusSizedNumber, PaillierModulusSizedNumber)>,
-    ) -> Result<()> {
-        let _ = self
-            .decryption_shares
-            .insert(party_id, message);
+        sender_id: PartyID,
+        message: SignMessage,
+    ) -> twopc_mpc_protocols::Result<()> {
+        match message {
+            SignMessage::DecryptionShares(shares) => {
+                let _ = self.decryption_shares.insert(sender_id, shares);
+            }
+            SignMessage::StartIAFlow(failed_messages_indices, involved_parties) => {
+                if self.failed_messages_indices.is_none() {
+                    self.failed_messages_indices = Some(failed_messages_indices.clone());
+                }
+                if self.involved_parties.is_none() {
+                    self.involved_parties = Some(involved_parties.clone());
+                }
+            }
+            SignMessage::Proofs((proofs)) => {
+                self.insert_proofs(sender_id, proofs.clone());
+            }
+        }
         Ok(())
     }
 
+    pub(crate) fn insert_proofs(&mut self, party_id: PartyID, new_proofs: Vec<PartialDecryptionProof>) {
+        if self.clone().involved_parties.unwrap().contains(&party_id) {
+            self.proofs.get_or_insert(HashMap::new()).insert(party_id, new_proofs);
+        }
+    }
+
     pub(crate) fn ready_for_complete_first_round(&self, round: &SignRound) -> bool {
         match round {
-            SignRound::FirstRound { .. } if self.decryption_shares.len() == self.parties.len() && self.party_id == self.aggregator_party_id => true,
-            _ => false
+            SignRound::FirstRound { .. } => {
+                self.received_all_decryption_shares() && self.party_id == self.aggregator_party_id
+            }
+            _ => false,
+        }
+    }
+
+    pub(crate) fn received_all_decryption_shares(&self) -> bool {
+        self.decryption_shares.len() == self.parties.len()
+    }
+
+    pub(crate) fn should_identify_malicious_actors(&self) -> bool {
+        if let Some(proofs) = self.clone().proofs {
+            let threshold: usize = self.tiresias_public_parameters.threshold.into();
+            return proofs.len() == threshold && self.received_all_decryption_shares();
         }
+        return false;
```

Comment:
We should handle the case that someone didn't send a proof after some time (i.e. Narwhal rounds) and report it as malicious. If you don't know how to do this, at least add a `todo` comment for this

Reply:
AFAIU we should not deal with it right now, as Tomer's team is working on it. Am I correct? @omersadika

Reply:
Even if so, we should add a `todo` comment for that in the right place

Reply:
I don't know where is the right place..

Reply:
That's correct. Add todo